weasyl/searchtag.py

```diff
@@ -217,3 +254,233 @@ def tag_history(submitid):
+    if userid not in staff.DIRECTORS:
+        raise WeasylError("InsufficientPermissions")
+
+    existing = d.engine.execute("""
+        SELECT tagid FROM searchmapglobalblacklist
+    """).fetchall()
+
+    # Retrieve tag titles and tagid pairs, for new and existing tags
+    query = add_and_get_searchtags(tags)
+
+    existing_tagids = {t.tagid for t in existing}
+    entered_tagids = {t.tagid for t in query}
+
+    # Assign added and removed
+    added = entered_tagids - existing_tagids
+    removed = existing_tagids - entered_tagids
+
+    if added:
+        d.engine.execute("""
+            INSERT INTO searchmapglobalblacklist (tagid, userid)
+                SELECT tag, %(uid)s
+                FROM UNNEST (%(added)s) AS tag
+        """, uid=userid, added=list(added))
+
+    if removed:
+        d.engine.execute("""
+            DELETE FROM searchmapglobalblacklist
+            WHERE tagid = ANY (%(removed)s)
+        """, removed=list(removed))
+
+    # Clear the global STBL cache if any changes were made
+    if added or removed:
+        query_global_blacklisted_tags.invalidate()
+
+
+def get_user_searchtag_blacklist(userid):
+    """
+    Retrieves a list of tags on the user searchtag blacklist for friendly display to the user.
+
+    Parameters:
+        userid: The userid of the user requesting the list of tags.
+
+    Returns:
+        A list of blacklisted searchtag titles which were set by ``userid``.
+    """
+    query = d.engine.execute("""
+        SELECT st.title
+        FROM searchmapuserblacklist
+        INNER JOIN searchtag AS st USING (tagid)
+        WHERE userid = %(userid)s
+        ORDER BY st.title
+    """, userid=userid).fetchall()
+    tags = [tag.title for tag in query]
+    return tags
+
+
+def get_global_searchtag_blacklist(userid):
+    """
+    Retrieves a list of tags on the global searchtag blacklist for friendly display to the director.
+
+    Parameters:
+        userid: The userid of the user requesting the list of tags.
+
+    Returns:
+        A list of globally blacklisted searchtag titles and the name of the user which added it.
+    """
+    # Only directors can view the global blacklist; sanity check against the @director_only decorator
+    if userid not in staff.DIRECTORS:
+        raise WeasylError("InsufficientPermissions")
+
+    query = d.engine.execute("""
+        SELECT st.title, lo.login_name
+        FROM searchmapglobalblacklist
+        INNER JOIN searchtag AS st USING (tagid)
+        INNER JOIN login AS lo USING (userid)
+        ORDER BY st.title
+    """).fetchall()
+    return query
+
+
+@region.cache_on_arguments()
+def query_user_blacklisted_tags(ownerid):
+    """
+    Gets and returns blacklisted searchtag blacklist tags for both user tags.
+
+    Parameters:
+        ownerid: The userid of the user who owns the content tags are being added to.
+
+    Returns:
+        user_blacklist_query: User STBL tag titles.
```

Comment:
The name `user_blacklist_query` doesn’t add any information here. (It’s not a query, either.)